charts/vineyard/templates/master/statefulset.yaml

```diff
@@ -0,0 +1,126 @@
+{{- $peerPort := int .Values.master.ports.peer }}
+{{- $name := include "vineyard.name" . }}
+{{- $fullName := include "vineyard.fullname" . }}
+{{- $etcdFullname := (printf "%s-%s" $fullName "master") }}
+{{- $chart := include "vineyard.chart" . }}
+{{- $releaseNamespace := .Release.Namespace }}
+{{- $etcdServiceName := (printf "%s-%s" $fullName "master-svc") }}
+
+{{- if not .Values.master.externalEndpoint }}
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ $fullName }}-master
+  labels:
+    app: {{ $name }}
+    chart: {{ $chart }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+    role: vineyard-master
+  {{- if .Values.owner.enabled }}
+  ownerReferences:
+  - apiVersion: {{ .Values.owner.apiVersion }}
+    blockOwnerDeletion: {{ .Values.owner.blockOwnerDeletion }}
+    controller: {{ .Values.owner.controller }}
+    kind: {{ .Values.owner.kind }}
+    name: {{ .Values.owner.name }}
+    uid: {{ .Values.owner.uid }}
+  {{- end }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ $name }}
+      chart: {{ $chart }}
+      release: {{ .Release.Name }}
+      role: vineyard-master
+  serviceName: {{ $fullName }}-master-svc
+  replicas: {{ $replicas }}
+  podManagementPolicy: Parallel
+  template:
+    metadata:
+      annotations:
+        sidecar.istio.io/inject: "false"
+      labels:
+        app: {{ $name }}
+        chart: {{ $chart }}
+        release: {{ .Release.Name }}
+        heritage: {{ .Release.Service }}
+        role: vineyard-master
+    spec:
+      {{- if .Values.master.nodeSelector }}
+      nodeSelector:
+{{ toYaml .Values.master.nodeSelector | trim | indent 8  }}
+      {{- end }}
+      containers:
+        - name: vineyard-master
+          image: {{ .Values.master.image }}:{{ .Values.master.imageTag }}
+          imagePullPolicy: {{ .Values.master.imagePullPolicy }}
+          securityContext:
+            runAsUser: 1001
+            runAsNonRoot: true
+            privileged: false
+          command:
+            - etcd
+            - --name=$(POD_NAME)
+            - --initial-advertise-peer-urls=http://$(POD_NAME).{{ $etcdServiceName }}.{{ .Release.Namespace }}.svc.cluster.local:{{ $peerPort }}
+            - --listen-peer-urls=http://0.0.0.0:{{ $peerPort }}
+            - --advertise-client-urls=http://$(POD_NAME).{{ $etcdServiceName }}.{{ .Release.Namespace }}.svc.cluster.local:{{ $clientPort }}
+            - --listen-client-urls=http://0.0.0.0:{{ $clientPort }}
+            - --initial-cluster-token=my-etcd-cluster
+            - --initial-cluster-state=new
+            {{- $initialCluster := list }}
+            {{- range $e, $i := until $replicas }}
+            {{- $initialCluster = append $initialCluster (printf "%s-%d=http://%s-%d.%s.%s.svc.cluster.local:%d" $etcdFullname $i $etcdFullname $i $etcdServiceName $releaseNamespace $peerPort) }}
+            {{- end }}
+            - --initial-cluster={{ join "," $initialCluster }}
```

Comment:
Is it possible to reuse the helm func `{{ include "vineyard.master.endpoint" . }} `here?

Reply:
Yes, `"vineyard.master.endpoint"` use the client port, but we need the peer port here. We could make the helm func suitable for the master port and the client port.